jmxtrans-core/src/main/java/com/googlecode/jmxtrans/model/JmxResultProcessor.java

```diff
@@ -58,132 +58,137 @@ public JmxResultProcessor(Query query, ObjectInstance objectInstance, List<Attri
-			processTabularDataSupport(accumulator, attribute.getName(), tds);
-			accumulator.add(r);
-		}  else if (value instanceof Map) {
-			Result r = getNewResultObject(attribute.getName(), convertKeysToString((Map<Object, Object>) value));
-			accumulator.add(r);
-		} else {
-			Map<String, Object> values = newHashMap();
-			values.put(attribute.getName(), value);
-			Result r = getNewResultObject(attribute.getName(), values);
-			accumulator.add(r);
+			addNew(attributeName, values.build());
 		}
-	}
 
-	private <K, V> ImmutableMap<String, V> convertKeysToString(Map<K, V> value) {
-		ImmutableMap.Builder<String, V> values = ImmutableMap.builder();
-		for (Map.Entry<K, V> entry : value.entrySet()) {
-			values.put(entry.getKey().toString(), entry.getValue());
+		/**
+		 * Add results from a value of type objet name array.
+		 */
+		private void add(String attributeName, ObjectName[] objs) {
+			ImmutableMap.Builder<String, Object> values = ImmutableMap.builder();
+			for (ObjectName obj : objs) {
+				values.put(obj.getCanonicalName(), obj.getKeyPropertyListString());
+			}
+			addNew(attributeName, values.build());
 		}
-		return values.build();
-	}
 
-	/**
-	 * Populates the Result objects. This is a recursive function. Query
-	 * contains the keys that we want to get the values of.
-	 */
-	private void getResult(Builder<Result> accumulator, String attributeName, CompositeData cds) {
-		CompositeType t = cds.getCompositeType();
-
-		Map<String, Object> values = newHashMap();
-
-		Set<String> keys = t.keySet();
-		for (String key : keys) {
-			Object value = cds.get(key);
-			if (value instanceof TabularDataSupport) {
-				TabularDataSupport tds = (TabularDataSupport) value;
-				processTabularDataSupport(accumulator, attributeName + "." + key, tds);
-				// continue because we added tabular contents within above, but need primitives at this level
-			} else if (value instanceof CompositeDataSupport) {
-				// now recursively go through everything.
-				CompositeDataSupport cds2 = (CompositeDataSupport) value;
-				getResult(accumulator, attributeName, cds2);
-				return; // because we don't want to add to the list yet.
-			} else {
-				values.put(key, value);
+		/**
+		 * Add results from a value of type composite data.
+		 * This is a recursive function.
+		 */
+		private void add(String attributeName, CompositeData cds) {
+			CompositeType t = cds.getCompositeType();
+
+			Map<String, Object> values = newHashMap();
+
+			Set<String> keys = t.keySet();
+			for (String key : keys) {
+				Object value = cds.get(key);
+				if (value instanceof TabularData) {
+					TabularData tds = (TabularData) value;
+					add(attributeName + "." + key, tds);
+					// continue because we added tabular contents within above, but need primitives at this level
+				} else if (value instanceof CompositeData) {
+					// now recursively go through everything.
+					CompositeData cds2 = (CompositeData) value;
+					add(attributeName, cds2);
```

Comment:
This part looks very strange:
* TabularData: add result with key "attributeName + data key"
* CompositeData: only add a result with the first element with key "attributeName"
* Simple type (I guess): add a result of type map
Personally, I would remove this "if tabular ... else if composite ... else ..." block and call the add(name, object) with name = name + key. This also means I would change the type of the "values" field in Result class from Map to Object, meaning each Result would contain a single and simple value.

What do you think?

Reply:
I think that each Result being a single object makes a lot of sense. If we manage to get each Result to be a single scalar value, this allow to simplify the OutputWriter quite a bit (I do suspect that a lot of them are broken and don't handle collections well).

Note for later: I think it would also make sense to abstract the metric name / path, instead of relying on strings. But that would also require modifying all existing OutputWriters...